src/main/scala/org/broadinstitute/hail/Utils.scala

```diff
@@ -568,6 +589,96 @@ class RichPairTraversableOnce[K, V](val t: TraversableOnce[(K, V)]) extends AnyV
   }
 }
 
+class RichPairIterator[K, V](val it: Iterator[(K, V)]) extends AnyVal {
+
+  def localKeySort[T](implicit ord: Ordering[T], kOrd: Ordering[K], ev: (K) => T): Iterator[(K, V)] = {
+    lazyKSorted(it)
+  }
+
+  /**
+    * Precondition: the iterator is T-sorted. Moreover, ev must be monotonic. We lazily K-sort each block of
+    * T-equivalent elements.
+    *
+    */
+  private def lazyKSorted[T](it: Iterator[(K, V)])
+    (implicit tOrd: Ordering[T], kOrd: Ordering[K], ev: (K) => T) : Iterator[(K, V)] =
+    if (it.isEmpty)
+      Iterator.empty
+    else {
+      val t = ev(it.next._1)
+      val (tEquivalents, greater) = it.span({ case (k, _) => tOrd.equiv(t, ev(k)) })
+      val kSorted = tEquivalents.toSeq.sortBy(_._1)
+      // NB: ++ is lazy in its second argument
+      kSorted.iterator ++ lazyKSorted(greater)
+    }
+
+  def sortedLeftJoinDistinct[W](other: Iterator[(K, W)])
+    (implicit ordering: Ordering[K]): Iterator[(K, (V, Option[W]))] = {
+    import ordering._
+
+    if (other.isEmpty)
+      it.map { case (k, v) => (k, (v, None)) }
+    else {
+      val (_1, _2) = other.next()
+      var k2 = _1
+      var v2 = _2
+
+      for ((k, v1) <- it) yield {
+        if (k2 > k)
+          (k, (v1, None))
+        else if (k2 == k)
+          (k, (v1, Some(v2)))
+        else {
+          while (k2 < k && other.hasNext) {
+            val (_1, _2) = other.next()
+            assert(_1 >= k2, "iterator was not sorted")
+            k2 = _1
+            v2 = _2
+          }
+          if (k2 == k)
+            (k, (v1, Some(v2)))
+          else
+            (k, (v1, None))
+        }
+      }
+    }
+  }
+
+  def sortedTransformedLeftJoinDistinct[T, V2](other: Iterator[(T, V2)])
```

Comment:
I wish there was a good abstraction over this and `sortedLeftJoinDistinct`, but I can't come up with something nice... @cseed ?